django_prices/models.py

```diff
@@ -1,91 +1,126 @@
+        setattr(cls, name, self)
+
+    def clean(self, value, model_instance):
+        # Shortcircut clean() because Django calls it on all fields with
+        # is_relation = False
         return value
 
-    def get_db_prep_save(self, value, connection):
-        value = self.get_prep_value(value)
-        db_value = connection.ops.adapt_decimalfield_value
-        return db_value(value, self.max_digits, self.decimal_places)
 
-    def value_to_string(self, obj):
-        value = self.value_from_object(obj)
+class MoneyField(NonDatabaseFieldBase):
+
+    description = (
+        "A field that combines an amount of money and currency code into Money"
+        "It allows to store prices with different currencies in one database."
+    )
+
+    def __init__(
+        self,
+        amount_field="price_amount",
+        currency_field="price_currency",
+        verbose_name=None,
+        **kwargs
+    ):
+        super(MoneyField, self).__init__()
+        self.amount_field = amount_field
+        self.currency_field = currency_field
+
+    def __str__(self):
+        return "MoneyField(amount_field=%s, currency_field=%s)" % (
+            self.amount_field,
+            self.currency_field,
+        )
+
+    def __get__(self, instance, cls=None):
+        if instance is None:
+            return self
+
+        amount = getattr(instance, self.amount_field)
+        currency = getattr(instance, self.currency_field)
+        if amount and currency:
+            return Money(amount, currency)
+        return self.get_default()
+
+    def __set__(self, instance, value):
+        amount = None
+        currency = None
         if value is not None:
-            return value
-        return super(MoneyField, self).value_to_string(value)
+            amount = value.amount
+            currency = value.currency
+        setattr(instance, self.amount_field, amount)
+        setattr(instance, self.currency_field, currency)
 
     def formfield(self, **kwargs):
-        defaults = {"currency": self.currency, "form_class": forms.MoneyField}
-        defaults.update(kwargs)
-        return super(MoneyField, self).formfield(**defaults)
+        available_currencies = []
+        if hasattr(self, "model"):
+            available_currencies = self.model._meta.get_field(
+                self.currency_field
+            ).choices
+        return forms.MoneyField(available_currencies=available_currencies)
 
     def get_default(self):
-        default = super(MoneyField, self).get_default()
-        return self.to_python(default)
+        default_currency = ""
+        default_amount = Decimal(0)
+        if hasattr(self, "model"):
+            default_currency = self.model._meta.get_field(
+                self.currency_field
+            ).get_default()
+            d = self.model._meta.get_field(self.amount_field).get_default()
+            default_amount = self.model._meta.get_field(self.amount_field).to_python(d)
```

Comment:
Why do we execute additional logic here? Shouldn't `get_default()` be enough?

Reply:
Thanks for finding that, it was some artifact from my previous version or experiment, I've removed it.